File: task-stack/stack.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
	"time"
)

type Task struct {
	Text        string     `json:"text"`
	LastCurrent *time.Time `json:"last_current"`
}

func stackPath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".task-stack.json")
}

func Load() ([]Task, error) {
	data, err := os.ReadFile(stackPath())
	if os.IsNotExist(err) {
		return []Task{}, nil
	}
	if err != nil {
		return nil, err
	}
	var tasks []Task
	if err := json.Unmarshal(data, &tasks); err != nil {
		return nil, err
	}
	return tasks, nil
}

func Save(tasks []Task) error {
	data, err := json.MarshalIndent(tasks, "", "  ")
	if err != nil {
		return err
	}
	tmp := stackPath() + ".tmp"
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		return err
	}
	return os.Rename(tmp, stackPath())
}

func Push(tasks []Task, text string) []Task {
	now := time.Now().UTC()
	return append([]Task{{Text: text, LastCurrent: &now}}, tasks...)
}

func Pop(tasks []Task) ([]Task, *Task) {
	if len(tasks) == 0 {
		return tasks, nil
	}
	t := tasks[0]
	return tasks[1:], &t
}

func Promote(tasks []Task, idx int) []Task {
	if idx <= 0 || idx >= len(tasks) {
		return tasks
	}
	now := time.Now().UTC()
	promoted := tasks[idx]
	promoted.LastCurrent = &now
	result := make([]Task, 0, len(tasks))
	result = append(result, promoted)
	result = append(result, tasks[:idx]...)
	return append(result, tasks[idx+1:]...)
}

func Remove(tasks []Task, idx int) []Task {
	if idx < 0 || idx >= len(tasks) {
		return tasks
	}
	result := make([]Task, 0, len(tasks)-1)
	result = append(result, tasks[:idx]...)
	return append(result, tasks[idx+1:]...)
}

func FormatTimestamp(t *time.Time, now time.Time) string {
	if t == nil {
		return "—"
	}
	l := t.Local()
	n := now.Local()
	switch {
	case l.Year() != n.Year():
		return l.Format("2006-01-02 15:04")
	case l.Month() != n.Month():
		return l.Format("01-02 15:04")
	case l.Day() != n.Day():
		return l.Format("02 15:04")
	case l.Hour() != n.Hour():
		return l.Format("15:04")
	default:
		return l.Format("04")
	}
}

File: task-stack/src/task_stack/stack.py
from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path

STACK_FILE = Path.home() / ".task-stack.json"
_TMP = STACK_FILE.with_suffix(".json.tmp")


@dataclass
class Task:
    text: str
    last_current: datetime | None = field(default=None)

    def to_dict(self) -> dict:
        return {
            "text": self.text,
            "last_current": self.last_current.isoformat() if self.last_current else None,
        }

    @staticmethod
    def from_dict(d: dict) -> "Task":
        lc = d.get("last_current")
        return Task(
            text=d["text"],
            last_current=datetime.fromisoformat(lc) if lc else None,
        )


def load() -> list[Task]:
    if not STACK_FILE.exists():
        return []
    try:
        data = json.loads(STACK_FILE.read_text())
        return [Task.from_dict(item) for item in data]
    except Exception:
        return []


def save(tasks: list[Task]) -> None:
    data = json.dumps([t.to_dict() for t in tasks], indent=2)
    _TMP.write_text(data)
    os.replace(_TMP, STACK_FILE)


def push(text: str) -> list[Task]:
    tasks = load()
    task = Task(text=text.strip(), last_current=datetime.now(tz=timezone.utc))
    tasks.insert(0, task)
    save(tasks)
    return tasks


def pop() -> tuple[Task | None, list[Task]]:
    tasks = load()
    if not tasks:
        return None, []
    removed = tasks.pop(0)
    if tasks:
        tasks[0].last_current = datetime.now(tz=timezone.utc)
    save(tasks)
    return removed, tasks


def reorder(from_idx: int, to_idx: int) -> list[Task]:
    tasks = load()
    if from_idx == to_idx or not (0 <= from_idx < len(tasks)) or not (0 <= to_idx < len(tasks)):
        return tasks
    task = tasks.pop(from_idx)
    tasks.insert(to_idx, task)
    if to_idx == 0:
        tasks[0].last_current = datetime.now(tz=timezone.utc)
    save(tasks)
    return tasks


def promote(idx: int) -> list[Task]:
    return reorder(idx, 0)


def remove(idx: int) -> list[Task]:
    tasks = load()
    if not (0 <= idx < len(tasks)):
        return tasks
    tasks.pop(idx)
    if tasks and idx == 0:
        tasks[0].last_current = datetime.now(tz=timezone.utc)
    save(tasks)
    return tasks


def format_timestamp(dt: datetime | None, now: datetime | None = None) -> str:
    if dt is None:
        return "—"
    if now is None:
        now = datetime.now(tz=timezone.utc)
    # Normalize both to UTC-aware for comparison
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    if now.tzinfo is None:
        now = now.replace(tzinfo=timezone.utc)

    if dt.year != now.year:
        return dt.strftime("%Y-%m-%d %H:%M")
    if dt.month != now.month:
        return dt.strftime("%m-%d %H:%M")
    if dt.day != now.day:
        return dt.strftime("%d %H:%M")
    if dt.hour != now.hour:
        return dt.strftime("%H:%M")
    return dt.strftime("%M")
